source/scpp/src/util/Math.cpp

```diff
@@ -17,20 +22,135 @@ rand_fraction()
     return uniformFractionDistribution(gRandomEngine);
 }
 
-size_t
-rand_pareto(float alpha, size_t max)
-{
-    // from http://www.pamvotis.org/vassis/RandGen.htm
-    float f =
-        static_cast<float>(1) /
-        static_cast<float>(pow(rand_fraction(), static_cast<float>(1) / alpha));
-    // modified into a truncated pareto
-    return static_cast<size_t>(f * max - 1) % max;
-}
-
 bool
 rand_flip()
 {
     return (gRandomEngine() & 1);
 }
+
+double
+closest_cluster(double p, std::set<double> const& centers)
+{
+    auto bestCenter = std::numeric_limits<double>::max();
+    auto currDist = std::numeric_limits<double>::max();
+    for (auto const& c : centers)
+    {
+        auto newDist = std::fabs(c - p);
+        if (newDist < currDist)
+        {
+            bestCenter = c;
+            currDist = newDist;
+        }
+        else
+        {
+            break;
+        }
+    }
+
+    return bestCenter;
+}
+
+std::set<double>
+k_meansPP(std::vector<double> const& points, uint32_t k)
+{
+    if (k == 0)
+    {
+        throw std::runtime_error("k_means: k must be positive");
+    }
+
+    if (points.size() < k)
+    {
+        return std::set<double>(points.begin(), points.end());
+    }
+
+    std::set<double> centroids;
+
+    auto backlog = points;
+
+    auto moveIndexToCentroid = [&](size_t index) {
+        // @BUG: undefined symbol printAssertFailureAndThrow()
+        // releaseAssertOrThrow(index < backlog.size());
```

Comment:
This can be uncommented if we add `GlobalChecks.o` to `dub.json`.
But I'd like to either get rid of it (and the `Backtace.cpp` dependency) or have it implemented in D.